core/src/execution/datafusion/expressions/cast.rs

```diff
@@ -1208,6 +1277,260 @@ fn do_cast_string_to_int<
+                            Some(v) => {
+                                exponent = v;
+                                continue;
+                            }
+                        }
+                    }
+                    if !b.is_ascii_digit() {
+                        return None;
+                    }
+                    fractionals += 1;
+                    digits += 1;
+                    result = result.mul_wrapping(base);
+                    result = result.add_wrapping(T::Native::usize_as((b - b'0') as usize));
+                }
+
+                // +00. is a valid decimal value in Spark but +. is invalid.
+                if digits == 0 && leading_zeros == 0 {
+                    return None;
+                }
+            }
+            b'e' | b'E' => match parse_exponent(
+                &mut bs,
+                &mut digits,
+                &mut fractionals,
+                &mut leading_zeros,
+                &mut has_exponent,
+                &mut has_negative_exp,
+            ) {
+                None => {
+                    return None;
+                }
+                Some(v) => {
+                    exponent = v;
+                    continue;
+                }
+            },
+            _ => {
+                return None;
+            }
+        }
+    }
+
+    result = adjust_decimal_scale::<T>(
+        result,
+        precision,
+        scale,
+        base,
+        exponent,
+        fractionals,
+        digits,
+        has_exponent,
+        has_negative_exp,
+    )?;
+
+    return Some(if negative {
+        result.neg_wrapping()
+    } else {
+        result
+    });
+}
+
+fn parse_exponent(
+    bs: &mut std::slice::Iter<u8>,
+    digits: &mut u8,
+    fractionals: &mut i8,
+    leading_zeros: &mut i8,
+    has_exponent: &mut bool,
+    has_negative_exp: &mut bool,
+) -> Option<i8> {
+    if *digits == 0 && *leading_zeros == 0 {
+        return None;
+    }
+
+    let mut exponent: i8 = 0;
+    let mut has_exp_digits = false;
+    *has_exponent = true;
+
+    for b in bs.by_ref() {
+        if *b == b'-' {
+            if *has_negative_exp || has_exp_digits {
+                return None;
+            }
+            *has_negative_exp = true;
+            continue;
+        } else if *b == b'+' {
+            if has_exp_digits {
+                return None;
+            }
+            continue;
+        }
+        if !b.is_ascii_digit() {
+            return None;
+        }
+        exponent = exponent.checked_mul(10)?.checked_add((b - b'0') as i8)?;
+        has_exp_digits = true;
+    }
+
+    if !has_exp_digits {
+        return None;
+    }
+    if !*has_negative_exp {
+        let cur_fractional = *fractionals as i8;
+        *fractionals = max(0, *fractionals - exponent);
+        exponent = max(0, exponent - cur_fractional);
+    } else {
+        *fractionals += exponent;
+    }
+    *digits += exponent.abs() as u8;
+
+    return Some(exponent);
```

Comment:
Same comment re unnecessary `return`

```suggestion
    Some(exponent)
```